keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -12,124 +11,477 @@ approvers:
+then the backend health will be checked using the parameters listed above. 
+
+### Status
+
+!!! TODO
+
+### Ingress class
+
+The `kubernetes.io/ingress.class` annotation is required for interoperation
+between various Ingress providers. As support for this annotation is universal,
+this concept should be promoted to an actual field.
+
+#### Proposal
+
+Promoting the annotation as is as an opaque string the most direct path, but
+precludes any future enhancements to the concept.
+
+An alternative is to create a new resource `IngressClass` to take its place.
+This resource will serve a couple of purposes:
+
+* Define the set of valid classes available to the user. Gives operators control
+  over allowed classes.
+* Allow us to evolve the API to express concepts such a levels of service
+  associated with a given Ingress controller.
+
+Add a field to `ingress.spec`:
+```golang
+type IngressSpec struct {
+  ...
+  // Class references an IngressClass resource in kube-system. This is used
+  // by the cluster Ingress controllers to determine which controller operates
+  // on this resource.
+  Class string
+  ...
+}
+```
+
+```golang
+// IngressClass represents the class of the Ingress, referenced by the
+// ingress.spec.
+type IngressClass struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  Spec IngressClassSpec
+}
+
+type IngressClassSpec struct {
+  // This is currently empty.
+}
+```
+
+### Alternative backend types
+
+The Ingress resource is an L7 description of a composite set of services. It
+current supports Kubernetes Services as a backends. However, there are many use
+cases where a portion of the HTTP requests could be routed to a different kind
+of resource. For example, serving content from an object storage
+([S3][s3-backend], [GCS][gcs-backend]) is a commonly requested feature.
+
+At the same time, we do not expect to enumerate all possible backends that could
+arise, nor do we expect that naming of the resources will be uniform in schema,
+parameters etc. Similarly, many of the resources will be platform specific.
+
+#### Proposal
+
+Add a field to the `IngressBackend` struct with an object reference:
+
+```golang
+type IngressBackend struct {
+  // Specifies the name of the referenced service. The service must exist in
+  // the same namespace as the Ingress object. The API server will
+  // automatically populate the Resource field below.
+  // +optional
+  ServiceName string
+
+  // Specifies the port of the referenced service. If unspecfied and the
+  // ServiceName is non-empty, the Service must expose a single port.
+  // +optional
+  ServicePort intstr.IntOrString
+
+  // Resource is an ObjectRef to another Kubernetes resource in the namespace
+  // of the Ingress object. If Resource is a Service, then ServiceName will
+  // populated to match the Resource. If both ServiceName and Resource are
+  // specified, then they must reference the same Service object.
+  // +optional
+  Resource v1.TypedLocalObjectReference
```

Comment:
Can we use the same approach in [HTTPIngressPath](https://github.com/kubernetes/kubernetes/blob/7d75b73e1d72cef7a0f0e2804f8e0a582ccb6b61/staging/src/k8s.io/api/networking/v1beta1/types.go#L169)? This would allow exposing features available in the particular implementation removing the ambiguity of annotations.

Reply:
Do you really want a CRD for path strings? 

Also, that doesn't help with the round-trip requirement -- we have to be able to go from v1beta1 -> v1 -> v1beta1 and not lose information

Reply:
> Do you really want a CRD for path strings?

No, I want to remove annotations to configure "things" in the path.

> we have to be able to go from v1beta1 -> v1 -> v1beta1 and not lose information

Good point.